File: mvvm-java/domain/src/main/java/com/wasisto/githubuserfinder/domain/usecases/SearchUsersUseCase.java
package com.wasisto.githubuserfinder.domain.usecases;

import com.wasisto.githubuserfinder.common.Logger;
import com.wasisto.githubuserfinder.domain.interfaces.GithubDataSource;
import com.wasisto.githubuserfinder.domain.interfaces.SearchHistoryDataSource;
import com.wasisto.githubuserfinder.domain.models.SearchHistoryItem;
import com.wasisto.githubuserfinder.domain.models.SearchUsersResult;

public class SearchUsersUseCase {

    private GithubDataSource githubDataSource;

    private SearchHistoryDataSource searchHistoryDataSource;

    private Logger logger;

    public SearchUsersUseCase(GithubDataSource githubDataSource, SearchHistoryDataSource searchHistoryDataSource, Logger logger) {
        this.githubDataSource = githubDataSource;
        this.searchHistoryDataSource = searchHistoryDataSource;
        this.logger = logger;
    }

    public SearchUsersResult execute(String query) throws Exception {
        try {
            searchHistoryDataSource.add(
                    new SearchHistoryItem() {{
                        setQuery(query);
                    }}
            );
        } catch (Exception e) {
            logger.warn("An error occurred while adding a query to the search history", e);
        }
        return githubDataSource.searchUsers(query);
    }
}

File: mvp-kotlin/domain/src/main/java/com/wasisto/githubuserfinder/domain/usecases/SearchUsersUseCase.kt
package com.wasisto.githubuserfinder.domain.usecases

import com.wasisto.githubuserfinder.common.Logger
import com.wasisto.githubuserfinder.domain.interfaces.GithubDataSource
import com.wasisto.githubuserfinder.domain.interfaces.SearchHistoryDataSource
import com.wasisto.githubuserfinder.domain.models.SearchHistoryItem
import com.wasisto.githubuserfinder.domain.models.SearchUsersResult

class SearchUsersUseCase(
    private val githubDataSource: GithubDataSource,
    private val searchHistoryDataSource: SearchHistoryDataSource,
    private val logger: Logger
) {
    operator fun invoke(query: String): SearchUsersResult {
        try {
            searchHistoryDataSource.add(SearchHistoryItem(query = query))
            logger.info("Query added to the search history")
        } catch (e: Exception) {
            logger.warn("An error occurred while adding a query to the search history", e)
        }
        return githubDataSource.searchUsers(query)
    }
}